File: tools/cantp.py
from math import ceil
from construct import *
import struct

# CANTP: CAN transfer protocol adapted from ISO 15765-2
# https://en.wikipedia.org/wiki/ISO_15765-2
# https://stackoverflow.com/questions/34783297/can-for-large-packets-of-data-exchange-without-using-standard-protocols

CAN_FRAME_MAX_SIZE = 8

CANTP_MODE_STREAM = 1
CANTP_MODE_CHUNKS = 2
CANTP_MODE_CRC32  = 4

CANTP_BUFFER_SIZE = 32
CANTP_STREAM_TRANSFER_SIZE = 8

CANTPSingleFrame = Struct(
    "code" / Const(0, Hex(Int8ul)),
    "data0" / Hex(Int8ul),
    "data1" / Hex(Int8ul),
    "data2" / Hex(Int8ul),
    "data3" / Hex(Int8ul),
    "data4" / Hex(Int8ul),
    "data5" / Hex(Int8ul),
    "data6" / Hex(Int8ul),
)
assert(CANTPSingleFrame.sizeof() == CAN_FRAME_MAX_SIZE)

CANTPStreamFrame = Struct(
    "data" / Hex(Int64ul),
)
assert(CANTPStreamFrame.sizeof() == CAN_FRAME_MAX_SIZE)

CANTPChunkedFrame = Struct(
    "code" / Const(CANTP_MODE_CHUNKS, Hex(Int8ul)),
    "size" / Hex(Int8ul),
    "index" / Hex(Int16ul),
    "data" / Hex(Int32ul),
)
assert(CANTPChunkedFrame.sizeof() == CAN_FRAME_MAX_SIZE)

CANTPRequest = Struct(
    "magic" / Const(0xdead, Hex(Int16ul)),
    "mode" / Hex(Int8ul),
    "pad" / Const(0, Hex(Int8ul)),
    "size" / Hex(Int32ul),
)
assert(CANTPRequest.sizeof() == CAN_FRAME_MAX_SIZE)

CANTPChecksum = Struct(
    "mode" / Const(CANTP_MODE_CRC32, Hex(Int8ul)),
    "pad1" / Const(0, Hex(Int8ul)),
    "crc" / Hex(Int32ul),
    "pad2" / Const(0, Hex(Int16ul)),
)
assert(CANTPChecksum.sizeof() == CAN_FRAME_MAX_SIZE)

# https://en.wikipedia.org/wiki/ISO_15765-2

can_frames = []
buffer = [0] * CANTP_BUFFER_SIZE

def crc_update(data, prev):
    crc = prev ^ data
    idx = 0
    while (idx < 32):
        if (crc & 0x80000000): crc = ((crc << 1) ^ 0x04C11DB7) & 0xFFFFFFFF
        else: crc = (crc << 1) & 0xFFFFFFFF
        idx += 1
    return crc

def send_frame(data, cls):
    assert(len(data) <= CAN_FRAME_MAX_SIZE)
    #print(cls.parse(data))
    x = struct.unpack("<Q", data)
    print("CAN TX: 0x%016x" % (x))
    can_frames.append(cls.parse(data))

def cantp_send_request_frame(mode, data, size):
    assert(size >= 0)
    frame = CANTPRequest.build(dict(
        size=size,
        mode=mode,
    ))
    send_frame(frame, CANTPRequest)

def cantp_send_single_frame(data, size):
    assert(len(data) <= 7)
    if (len(data) < 7):
        data += [0] * (7 - len(data))
    data = CANTPSingleFrame.build(dict(
        data0 = data[0],
        data1 = data[1],
        data2 = data[2],
        data3 = data[3],
        data4 = data[4],
        data5 = data[5],
        data6 = data[6],
    ))
    send_frame(data, CANTPSingleFrame)

def cantp_send_chunked_frames(data, size):
    assert(size >= 8)
    count = int(ceil(size / 4))
    remaining = size
    bitpos = 0
    for n in range(count):
        s = min(remaining, 4)
        u32data = data[bitpos:bitpos+s]
        u32data = u32data[3] << 24 | u32data[2] << 16 | u32data[1] << 8 | u32data[0]
        frame = CANTPChunkedFrame.build(dict(code=2, size=s, index=n, data=u32data))
        send_frame(frame, CANTPChunkedFrame)
        remaining -= s
        bitpos += 4

def cantp_send_checksum(crc):
    frame = CANTPChecksum.build(dict(crc=crc))
    send_frame(frame, CANTPChecksum)

def cantp_send_stream_frames(data, size):
    assert(size >= 8)
    transfer_size = CANTP_STREAM_TRANSFER_SIZE  # 64 bits

    crc = 0xFFFFFFFF
    count = int(size // transfer_size)
    for n in range(count):
        u64data = data[n * transfer_size:(n + 1) * transfer_size]
        u64data1 = u64data[3] << 24 | u64data[2] << 16 | u64data[1] << 8 | u64data[0]
        u64data2 = u64data[7] << 24 | u64data[6] << 16 | u64data[5] << 8 | u64data[4]
        u64data = u64data2 << 32 | u64data1
        frame = CANTPStreamFrame.build(dict(data=u64data))
        send_frame(frame, CANTPStreamFrame)
        crc = crc_update(u64data1, crc)
        crc = crc_update(u64data2, crc)

    remaining = len(data) % transfer_size
    if (remaining): # send last frame
        u64data = 0
        for n in range(remaining):
            u64data |= data[count * transfer_size + n] << (n * 8)
        frame = CANTPStreamFrame.build(dict(data=u64data))
        send_frame(frame, CANTPStreamFrame)
        crc = crc_update(u64data & 0xffffffff, crc)
        crc = crc_update((u64data << 32) & 0xffffffff, crc)
    cantp_send_checksum(crc)

def cantp_send_data(mode, data, size):
    assert(len(data) == size)

    cantp_send_request_frame(mode, data, size)
    if (len(data) <= 7):
        cantp_send_single_frame(data, size) # take precedence
    elif (mode == CANTP_MODE_STREAM):
        cantp_send_stream_frames(data, size)
    elif (mode == CANTP_MODE_CHUNKS):
        cantp_send_chunked_frames(data, size)
    else:
        raise RuntimeError("wtf")

def main_tx():
    DATA_SIZE = 27
    data = [0] * DATA_SIZE
    for i in range(DATA_SIZE):
        data[i] = i
    size = len(data)
    cantp_send_data(CANTP_MODE_STREAM, data, size)

def cantp_receive_frame():
    global can_frames
    frame = can_frames[0]
    can_frames = can_frames[1:]
    return frame

def cantp_receieve_stream(size):
    if (size > CANTP_BUFFER_SIZE):
        raise ValueError("TODO error message")

    # TODO mutex lock
    crc = 0xFFFFFFFF
    count = int(ceil(size / CANTP_STREAM_TRANSFER_SIZE))
    for n in range(count):
        frame = cantp_receive_frame()
        buffer[n] = int(frame.data)
        data = int(frame.data)
        crc = crc_update(data & 0xffffffff, crc)
        crc = crc_update((data >> 32) & 0xffffffff, crc)

    frame = cantp_receive_frame() # crc
    assert(int(frame.mode) == CANTP_MODE_CRC32)
    assert(int(frame.crc) == crc)
    # TODO mutex unlock

    return buffer

def main_rx():
    frame = cantp_receive_frame()
    if (int(frame.magic) == 0xDEAD): # request
        if (int(frame.mode) == CANTP_MODE_STREAM):
            data = cantp_receieve_stream(int(frame.size))
        elif (int(frame.mode) == CANTP_MODE_CHUNKS):
            raise ValueError("TODO")
            #data = cantp_receieve_stream(int(frame.size))

main_tx()
main_rx()

File: tools/cantp.c

/**
 * @file cantp.c
 * @author Eileen Yoon (eyn@purdue.edu)
 * @brief
 * @version 0.1
 * @date 2024-11-18
 *
 * @copyright Copyright (c) 2024
 *
 * CAN Transport Protocol (for larger data packets) adapted from ISO 15765-2
 */

#include <assert.h>
#include <stdint.h>

#include <stdio.h>
//#include "common/phal_F4_F7/crc/crc.h"
#include "cantp.h"
#include "crc.h"

#define CAN_FRAME_MAX_SIZE 8
#define CANTP_STREAM_TRANSFER_SIZE (CAN_FRAME_MAX_SIZE)

#define CANTP_FRAME_DIRECTION_TX 0
#define CANTP_FRAME_DIRECTION_RX 1

static volatile uint32_t CRC32_MASTER = 0;

static uint64_t CANTP_FRAMES[256];
static int CANTP_RX_FRAME_INDEX1 = 0;
static int CANTP_RX_FRAME_INDEX2 = 0;

struct cantp_request_frame {
    uint16_t magic;
    uint8_t mode;
    uint8_t direction;
    uint32_t size;
} __attribute__((__packed__));
static_assert(sizeof(struct cantp_request_frame) == CAN_FRAME_MAX_SIZE);

struct cantp_checksum_frame {
    uint8_t mode;
    uint8_t pad1;
    uint32_t checksum;
    uint16_t pad2;
} __attribute__((__packed__));
static_assert(sizeof(struct cantp_checksum_frame) == CAN_FRAME_MAX_SIZE);

static int can_tx_send_u64(uint64_t data)
{
    printf("CAN TX: 0x%016lx\n", data);
    CANTP_FRAMES[CANTP_RX_FRAME_INDEX1++] = data;
    return 0;
}

static int cantp_send_request_frame(uint8_t mode, uint8_t *data, uint32_t size)
{
    struct cantp_request_frame frame = {
        .magic = 0xdead,
        .mode = mode,
        .direction = CANTP_FRAME_DIRECTION_TX,
        .size = size,
    };
    return can_tx_send_u64(*(uint64_t *)&frame);
}

static int cantp_send_checksum(uint32_t checksum)
{
    struct cantp_checksum_frame frame = {
        .mode = CANTP_MODE_CRC32,
        .pad1 = 0,
        .checksum = checksum,
        .pad2 = 0,
    };
    return can_tx_send_u64(*(uint64_t *)&frame);
}

static int cantp_send_single_frame(uint8_t *data, uint32_t size)
{
    uint64_t data64 = 0;
    for (int i = 0; i < size; i++)
        data64 |= ((uint64_t)data[i] << (i * 8)) & 0xff;
    can_tx_send_u64(data64);
}

void PHAL_reset_CRC32()
{
    CRC32_MASTER = 0xffffffff;
}

void PHAL_update_CRC32(uint32_t val)
{
    CRC32_MASTER = crc32_fast(CRC32_MASTER, val);
}

uint32_t PHAL_get_CRC32()
{
    return CRC32_MASTER;
}

static int cantp_send_stream_frames(uint8_t *data, uint32_t size)
{
    PHAL_reset_CRC32();

    uint64_t data64;
    int i;
    int count = size / CANTP_STREAM_TRANSFER_SIZE;
    for (i = 0; i < count; i++)
    {
        data64 = *(uint64_t *)(data + (i * CANTP_STREAM_TRANSFER_SIZE));
        can_tx_send_u64(data64);
        PHAL_update_CRC32(data64 & 0xffffffff);
        PHAL_update_CRC32((data64 >> 32) & 0xffffffff);
    }

    if (size % CANTP_STREAM_TRANSFER_SIZE) // send last frame
    {
        data64 = 0;
        for (i = 0; i < size % CANTP_STREAM_TRANSFER_SIZE; i++)
            data64 |= data[count * CANTP_STREAM_TRANSFER_SIZE + i] << (i * 8);
        can_tx_send_u64(data64);
        PHAL_update_CRC32(data64 & 0xffffffff);
        PHAL_update_CRC32((data64 >> 32) & 0xffffffff);
    }

    cantp_send_checksum(PHAL_get_CRC32());

    PHAL_reset_CRC32();

    return 0;
}

int cantp_send_data(uint8_t mode, uint8_t *data, uint32_t size)
{
    cantp_send_request_frame(CANTP_MODE_STREAM, data, size);
    if (size <= CANTP_STREAM_TRANSFER_SIZE) // single frame takes precedence
    {
        cantp_send_single_frame(data, size);
    }
    else if (mode == CANTP_MODE_STREAM)
    {
        cantp_send_stream_frames(data, size);
    }

    return 0;
}

// RX
uint8_t CANTP_RX_BUFFER[256];

static uint64_t cantp_receive_frame(void)
{
    uint64_t data = CANTP_FRAMES[CANTP_RX_FRAME_INDEX2++];
    printf("CAN RX: 0x%016lx\n", data);
    return data;
}

static uint32_t cantp_receive_checksum(void)
{
    uint64_t data = cantp_receive_frame(); // get CRC
    struct cantp_checksum_frame *frame = (struct cantp_checksum_frame *)(&data);
    if (frame->mode != CANTP_MODE_CRC32)
        return -1;
    return frame->checksum;
}

static int cantp_receive_stream_frames(uint8_t *buffer, uint32_t size)
{
    PHAL_reset_CRC32();

    uint64_t *buffer64 = (uint64_t *)buffer;
    uint64_t data64;
    int count = size / CANTP_STREAM_TRANSFER_SIZE;
    int i;

    // TODO mutex lock
    for (i = 0; i < count; i++)
    {
        data64 = cantp_receive_frame();
        buffer64[i] = data64;
        PHAL_update_CRC32(data64 & 0xffffffff);
        PHAL_update_CRC32((data64 >> 32) & 0xffffffff);
    }

    if (size % CANTP_STREAM_TRANSFER_SIZE) // send last frame
    {
        data64 = cantp_receive_frame();
        for (i = 0; i < size % CANTP_STREAM_TRANSFER_SIZE; i++)
        {
            buffer[count * CANTP_STREAM_TRANSFER_SIZE + i] = (data64 >> (i * 8)) & 0xff;
        }
        PHAL_update_CRC32(data64 & 0xffffffff);
        PHAL_update_CRC32((data64 >> 32) & 0xffffffff);
    }

    uint32_t checksum = cantp_receive_checksum();
    if (checksum != PHAL_get_CRC32())
    {
        printf("crc failed: 0x%08x vs 0x%08x\n", checksum, PHAL_get_CRC32());
        return -1;
    }

    PHAL_reset_CRC32();

    return 0;
}

static int cantp_receive_single_frame(uint8_t *buffer, uint32_t size)
{
    uint64_t data = cantp_receive_frame();
    for (int i = 0; i < size; i++)
    {
        buffer[i] = (data >> (i * 8)) & 0xff;
    }
}

int cantp_receive_data(void)
{
    uint64_t data;

    data = cantp_receive_frame();
    if ((uint16_t)(data & 0xffff) == 0xdead)
    {
        struct cantp_request_frame *frame = (struct cantp_request_frame *)(&data);

        if (frame->size <= CANTP_STREAM_TRANSFER_SIZE) // single frame takes precedence
        {
            cantp_receive_single_frame(CANTP_RX_BUFFER, frame->size);

        }
        else if (frame->mode == CANTP_MODE_STREAM)
        {
            cantp_receive_stream_frames(CANTP_RX_BUFFER, frame->size);
        }
        printf("buffer: 0x%x 0x%x 0x%x\n", CANTP_RX_BUFFER[0], CANTP_RX_BUFFER[1], CANTP_RX_BUFFER[2]);
    }

    return 0;
}

#define DATA_SIZE 15

int main(void)
{
    uint8_t data[DATA_SIZE];
    for (int i = 0; i < DATA_SIZE; i++)
    {
        data[i] = i;
    }
    cantp_send_data(1, data, DATA_SIZE);
    cantp_receive_data();

    return 0;
}
